pennylane_qulacs/qulacs_device.py

```diff
@@ -53,167 +93,215 @@ def hermitian(*args):
-            if len(par[0]) != 2 ** len(wires):
-                raise ValueError('Unitary matrix must be of shape (2**wires, 2**wires).')
-
-            unitary_gate = gate.DenseMatrix(wires, par[0])
-            self._circuit.add_gate(unitary_gate)
-        elif operation == 'Rot':
-            self._circuit.add_gate(gate.merge([
-                gate.RZ(wires[0], par[0]),
-                gate.RY(wires[0], par[1]),
-                gate.RZ(wires[0], par[2])
-            ]))
-        elif operation in ('CRZ', 'Toffoli', 'CSWAP'):
-            mapped_operation = self._operations_map[operation]
-            if callable(mapped_operation):
-                gate_matrix = mapped_operation(*par)
-            else:
-                gate_matrix = mapped_operation
-
-            dense_gate = gate.DenseMatrix(wires, gate_matrix)
-            self._circuit.add_gate(dense_gate)
+            self._state = QuantumStateGpu(self.num_wires)
         else:
-            mapped_operation = self._operations_map[operation]
-            self._circuit.add_gate(mapped_operation(*wires, *par))
+            self._state = QuantumState(self.num_wires)
 
-    @property
-    def state(self):
-        return self._state.get_vector()
-
-    def pre_measure(self):
-        self._circuit.update_quantum_state(self._state)
+        self._circuit = QuantumCircuit(self.num_wires)
 
-    def expval(self, observable, wires, par):
-        bra = self._state.copy()
+    def apply(self, operations, **kwargs):
+        rotations = kwargs.get("rotations", [])
+
+        self.apply_operations(operations)
+
+        # Rotating the state for measurement in the computational basis
+        self.apply_operations(rotations)
+
+    def apply_operations(self, operations):
+        """Apply the circuit operations to the state.
+
+        This method serves as an auxiliary method to :meth:`~.QulacsDevice.apply`.
+
+        Args:
+            operations (List[pennylane.Operation]): operations to be applied
+        """
+
+        for op in operations:
+            wires = op.wires
+            par = op.parameters
+
+            mapped_op = self._operation_map[op.name]
+            if op.name.endswith(".inv"):
+                # if an inverse variant of the operation exists
+                try:
+                    mapped_operation = getattr(gate, mapped_op.get_name() + "dag")
+                except AttributeError:
+                    # if the operation is hard-coded
+                    try:
+                        mapped_operation = np.conj(mapped_op).T
+                    except TypeError:
+                        # else redefine the operation as the inverse matrix
+                        mapped_operation = lambda *p: gate.DenseMatrix(
+                            wires, np.conj(
+                                mapped_op(*p).get_matrix()
+                            ).T
+                        )
+            else:
+                mapped_operation = mapped_op
```

Comment:
This adds the inverse functions in the following order:

1. If there's a build in hermitian function with the `dag` ending, use that one.
2. Check if the function in hard-coded (only `crz` for now) and just calculated the hermitian with numpy.
3. Otherwise, calculate the matrix representation of the gate, then the hermitian with numpy and lastly insert it into the `DenseMatrix` gate, and make it callable so that it can be used in the same way as a built in gate further down.

Reply:
This might be possible to improve upon. :slightly_smiling_face: